src/Microsoft.DotNet.Darc/src/DarcLib/Models/Darc/DependencyFlowGraph.cs

```diff
@@ -127,20 +127,206 @@ public void RemoveEdge(DependencyFlowEdge edge)
+                var successors = currentNode.IncomingEdges.Select(e => e.From);
+
+                // Compute the intersection of the dominators for the predecessors
+                foreach (var predNode in predecessors)
+                {
+                    if (newDom == null)
+                    {
+                        newDom = new HashSet<DependencyFlowNode>(dominators[predNode]);
+                    }
+                    else
+                    {
+                        newDom.IntersectWith(dominators[predNode]);
+                    }
+                }
+
+                if (newDom == null)
+                {
+                    newDom = new HashSet<DependencyFlowNode>();
+                }
+
+                // Add the current node
+                newDom.Add(currentNode);
+
+                // Compare to the existing dominator set for this node
+                if (!dominators[currentNode].SetEquals(newDom))
+                {
+                    dominators[currentNode] = newDom;
+
+                    // Queue all of the successors
+                    foreach (var succ in successors)
+                    {
+                        workList.Enqueue(succ);
+                    }
+                }
+            }
+
+            // Determine backedges
+            List<DependencyFlowEdge> toRemove = new List<DependencyFlowEdge>();
+            foreach (var edge in Edges)
+            {
+                if (dominators[edge.To].Contains(edge.From))
+                {
+                    edge.BackEdge = true;
+                }
+            }
+
+            // Remove the start node we created and links to it
+            foreach (var edge in startNode.IncomingEdges)
+            {
+                RemoveEdge(edge);
+            }
+
+            RemoveNode(startNode);
+        }
+
+        /// <summary>
+        ///     Calculate the longest build path time from each node in the flow graph
+        /// </summary>
+        /// <remarks>
+        ///     Starting with the set of root node (nodes with no outgoing edges), compute the
+        ///     longest build path time using a breadth first search
+        /// </remarks>
+        public void CalculateLongestBuildPaths()
+        {
+            List<DependencyFlowNode> roots = Nodes.Where(n => n.OutgoingEdges.Count == 0).ToList();
+
+            Queue<DependencyFlowNode> nodesToVisit = new Queue<DependencyFlowNode>();
+
+            foreach (var root in roots)
+            {
+                nodesToVisit.Enqueue(root);
+
+                while (nodesToVisit.Count > 0)
+                {
+                    DependencyFlowNode node = nodesToVisit.Dequeue();
+                    node.VisitedNodes.Add(node);
+
+                    foreach (var edge in node.IncomingEdges)
+                    {
+                        DependencyFlowNode child = edge.From;
+
+                        // Match based on repository and branch since we are updating 
```

Comment:
Since we are updating...?

Reply:
Oops. Probably left in some comments that don't make sense.